cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1870,195 +1731,301 @@ class S3FileSystem::Impl : public std::enable_shared_from_this<S3FileSystem::Imp
+      dir_infos.push_back(MakeDirectoryInfo(std::move(dirname)));
     }
-    return Status::OK();
+    return dir_infos;
   }
 
-  // A helper class for Walk and WalkAsync
-  struct FileInfoCollector {
-    FileInfoCollector(std::string bucket, std::string key, const FileSelector& select)
-        : bucket(std::move(bucket)),
-          key(std::move(key)),
-          allow_not_found(select.allow_not_found) {}
+  using FileInfoSink = PushGenerator<std::vector<FileInfo>>::Producer;
+
+  struct FileListerState {
+    FileInfoSink files_queue;
+    bool allow_not_found;
+    int max_recursion;
+    bool include_virtual;
+    S3Model::ListObjectsV2Request req;
+    io::IOContext io_context;
+    std::shared_ptr<Aws::S3::S3Client> client;
+    bool close_sink;
+    bool no_files_means_not_found;
+    std::unordered_set<std::string> directories;
+    bool empty = true;
+
+    FileListerState(PushGenerator<std::vector<FileInfo>>::Producer files_queue,
+                    FileSelector select, const std::string& bucket,
+                    const std::string& key, bool include_virtual,
+                    io::IOContext io_context, std::shared_ptr<Aws::S3::S3Client> client,
+                    bool close_sink)
+        : files_queue(std::move(files_queue)),
+          allow_not_found(select.allow_not_found),
+          max_recursion(select.max_recursion),
+          include_virtual(include_virtual),
+          io_context(io_context),
+          client(std::move(client)),
+          close_sink(close_sink),
+          no_files_means_not_found(!select.allow_not_found && !key.empty()) {
+      req.SetBucket(bucket);
+      req.SetMaxKeys(kListObjectsMaxKeys);
+      if (!key.empty()) {
+        req.SetPrefix(key + kSep);
+      }
+      if (!select.recursive) {
+        req.SetDelimiter(Aws::String() + kSep);
+      }
+    }
+
+    // The FileListerState is kept alive by the various FileListerTasks.  Once all the
+    // tasks are finished this will be destroyed and we can run some cleanup
+    ~FileListerState() {
+      // * If the bucket doesn't exist we will have already gotten an error from the
+      //     ListObjectsV2 call
+      // * If the key is empty, and the bucket exists, then there is
+      //     no way we can hit "not found"
+      // * If they key is not empty, then it's possible
+      //     that the file itself didn't exist and there
+      //     were not files under it.  In that case we will hit this if statement and
+      //     should treat this as a "not found" case.
+      if (empty && no_files_means_not_found) {
+        files_queue.Push(PathNotFound(req.GetBucket(), req.GetPrefix()));
+      }
+      if (close_sink) {
+        files_queue.Close();
+      }
+    }
 
-    Status Collect(const std::string& prefix, const S3Model::ListObjectsV2Result& result,
-                   std::vector<FileInfo>* out) {
-      // Walk "directories"
+    std::vector<std::string> GetNewDirectories(const std::string_view& path) {
```

Comment:
Please add a comment explaining what this does.

Reply:
Done.